plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -456,26 +458,29 @@ public DeltaLakeMetadata(
         this.allowManagedTableRename = allowManagedTableRename;
     }
 
-    public TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, long atVersion)
-    {
-        return getSnapshot(session, table, tableLocation, Optional.of(atVersion));
-    }
-
-    @VisibleForTesting
-    protected TableSnapshot getSnapshot(ConnectorSession session, SchemaTableName table, String tableLocation, Optional<Long> atVersion)
+    public TableSnapshot getSnapshot(
+            ConnectorSession session,
+            SchemaTableName table,
+            String tableLocation,
+            Optional<Long> atVersion,
+            Optional<ConnectorTableVersion> endVersion)
     {
+        TrinoFileSystem fileSystem = fileSystemFactory.create(session);
+        Optional<Long> endSnapshotVersion = endVersion.map(version -> getSnapshotIdFromVersion(fileSystem, tableLocation, version));
         try {
             if (atVersion.isEmpty()) {
                 atVersion = Optional.ofNullable(queriedVersions.get(table));
             }
             if (atVersion.isPresent()) {
                 long version = atVersion.get();
                 TableSnapshot snapshot = queriedSnapshots.get(new QueriedTable(table, version));
-                checkState(snapshot != null, "No previously loaded snapshot found for query %s, table %s [%s] at version %s", session.getQueryId(), table, tableLocation, version);
```

Comment:
This caching mechanism assumes that there is only one version table for each query.
Probably it should be adapted to time travel, to add also versions for source table verstions?